Ludox decision

danbryce opened the pull request:
- Bug fixes for execution
- examples of decision node in ludox protocol
- improvements to execution engine to allow graceful failure and return errors/warnings
- Change condition for FinalNode to be enabled to require any number (more than zero) tokens.

danbryce commented:
A large number of the changes are due to applying black formatting.  I plan to format everything after this gets merged.

danbryce commented:
@bbartley I'd like to rebase on this branch, so I'm wondering if you can have a look before I merge it.  Thanks!

bbartley commented on labop/__init__.py:
```diff
@@ -276,7 +276,7 @@ def show_libraries():
         show_library(lib)
 
-def get_primitive(doc: sbol3.Document, name: str):
+def get_primitive(doc: sbol3.Document, name: str, copy_to_doc: bool = True):
```
I can anticipate the `copy_to_doc` flag might be useful in some cases.  I'm curious, what was the use case that caused you to put it in?

bbartley commented on labop/container-ontology.ttl:
```diff
@@ -286,8 +274,7 @@ cont:width rdf:type owl:ObjectProperty ;
 ###  http://www.ontology-of-units-of-measure.org/resource/om-2/hasNumericalValue
-om:hasNumericalValue rdf:type owl:DatatypeProperty ;
-                     rdfs:range xsd:decimal .
+om:hasNumericalValue rdfs:range xsd:decimal .
```
If I am reading the diff correctly, it appears an important triple got deleted here:
```
om:hasNumericalValue rdf:type owl:DatatypeProperty
```
I think SBOLFactory will silently fail to generate the `hasNumericalValue` attribute if this is missing.

bbartley replied:
Sorry, disregard that previous comment, I just realized I am looking at container-ontology, not labop!

bbartley commented on labop/execution_engine.py:
```diff
@@ -348,34 +429,59 @@ def protocol_execution_aggregate_child_materials(self):
 
         if isinstance(source, uml.Pin):
-            src_parameter = source.get_parent().pin_parameter(source.name).property_value
+            src_parameter = (
```
Are the parentheses here for line continuation recommended by PEP 8 style guide, or is it a personal preference?  I don't have a problem with it, just trying to get a feel for your stylistic preferences.

bbartley commented on labop_convert/behavior_specialization.py:
```diff
@@ -51,113 +50,110 @@ def on_begin(self, execution: labop.ProtocolExecution):
+            "https://bioprotocols.org/labop/primitives/liquid_handling/TransferByMap": self.handle,
+            "http://bioprotocols.org/labop#Protocol": self.handle,
         }
-
-
-def validate_spec_query(query: str) -> "tyto.URI":
-    if type(query) is tyto.URI:
-        return query
-
-    if '#' in query:
-        # Query is assumed to be a URI
-        tokens = query.split('#')
-        if len(tokens) > 2 or tokens[0] != ContO.uri:
-            raise ValueError(f"Cannot resolve container specification '{query}'. The query is not a valid URI")
-        return tyto.URI(query, ContO)
-
-    # Query is assumed to be a qname
-    if ':' in query:
-        tokens = query.split(':')
```
Removing this is going to break the labware inference example as well.

bbartley commented on labop_convert/behavior_specialization.py:
```diff
@@ -51,113 +50,110 @@ def on_begin(self, execution: labop.ProtocolExecution):
-        if container_uri.is_instance():
-            possible_container_types = [container_uri]
-        else:
-            possible_container_types = container_uri.get_instances()
```
This code which has been reverted will break the labware inference example.  

The code here first attempts to resolve a labware reference using a remote or dev ontology server, but, if it can't find one, will default to using the local, static copy of the container-ontology.ttl.

bbartley reviewed (changes requested):
Thanks for putting the pre-commit hooks in to keep the codebase stylistically clean.  We should discuss whether we want to include a pycodestyle check as well.  Tom Mitchell devised a pretty clever pre-commit check for pySBOL3 that encourages clean-up of  existing code without being an obstacle to new development.

I noticed a lot of single quotes got changed to double quotes.  We should standardize on a convention.  I tend to prefer single quotes, but the choice is arbitrary and I'm not dogmatic about it.

The biggest problem that I noticed here is that there is a reversion of some big chunks of code necessary for the labware inference example in the JETC paper.  These appear to be in opentrons_specialization.py and behavior_specialization.py.  I tried to comment where I saw them, but may not have got it all.  Let me know how you want to proceed and if you need help resolving it.

danbryce commented:
The primitives in primitive_execution.py that define a compute_output() function were being added to the protocol doc.  I.e., we call get_primitive() to get the class, and then attach compute_output().  I didn’t want to change the default behavior of get_primitive(), which is to add the primitive to the doc.  I believe its used when importing libraries into a doc for the library.

> On Mar 4, 2023, at 10:09 AM, bbartley ***@***.***> wrote:
> 
> 
> @bbartley commented on this pull request.
> 
> In labop/__init__.py:
> 
> > @@ -276,7 +276,7 @@ def show_libraries():
>      for lib in labop.loaded_libraries.keys():
>          show_library(lib)
>  
> -def get_primitive(doc: sbol3.Document, name: str):
> +def get_primitive(doc: sbol3.Document, name: str, copy_to_doc: bool = True):
> 
> I can anticipate the copy_to_doc flag might be useful in some cases. I'm curious, what was the use case that caused you to put it in?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.
>

danbryce commented:
I had merged your jetc-revert branch, which had this change.  Can you say more about that branch and whether we need it or not?

> On Mar 4, 2023, at 10:11 AM, bbartley ***@***.***> wrote:
> 
> 
> @bbartley commented on this pull request.
> 
> In labop/container-ontology.ttl:
> 
> > @@ -286,8 +274,7 @@ cont:width rdf:type owl:ObjectProperty ;
>  #################################################################
>  
>  ###  http://www.ontology-of-units-of-measure.org/resource/om-2/hasNumericalValue
> -om:hasNumericalValue rdf:type owl:DatatypeProperty ;
> -                     rdfs:range xsd:decimal .
> +om:hasNumericalValue rdfs:range xsd:decimal .
> 
> If I am reading the diff correctly, it appears an important triple got deleted here:
> 
> om:hasNumericalValue rdf:type owl:DatatypeProperty
> 
> I think SBOLFactory will silently fail to generate the hasNumericalValue attribute if this is missing.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.
>

danbryce commented:
I applied the black formatter, which does this.  I plan to make this a global formatting change once we merge our outstanding branches.

> On Mar 4, 2023, at 10:27 AM, bbartley ***@***.***> wrote:
> 
> 
> @bbartley commented on this pull request.
> 
> In labop/execution_engine.py:
> 
> >  
>          if isinstance(source, uml.Pin):
> -            src_parameter = source.get_parent().pin_parameter(source.name).property_value
> +            src_parameter = (
> 
> Are the parentheses here for line continuation recommended by PEP 8 style guide, or is it a personal preference? I don't have a problem with it, just trying to get a feel for your stylistic preferences.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.
>

danbryce commented:
This was also part of the jetc-revision branch.

> On Mar 4, 2023, at 10:49 AM, bbartley ***@***.***> wrote:
> 
> 
> @bbartley commented on this pull request.
> 
> In labop_convert/behavior_specialization.py:
> 
> > -
> -
> -def validate_spec_query(query: str) -> "tyto.URI":
> -    if type(query) is tyto.URI:
> -        return query
> -
> -    if '#' in query:
> -        # Query is assumed to be a URI
> -        tokens = query.split('#')
> -        if len(tokens) > 2 or tokens[0] != ContO.uri:
> -            raise ValueError(f"Cannot resolve container specification '{query}'. The query is not a valid URI")
> -        return tyto.URI(query, ContO)
> -
> -    # Query is assumed to be a qname
> -    if ':' in query:
> -        tokens = query.split(':')
> 
> Removing this is going to break the labware inference example as well.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.
>

bbartley commented:
I don't think we need the `jetc-revert` branch.  That branch should be pruned.  I can't recall the problem I was trying to fix, but I think that I ended up resolving it without reverting.  
Looking back, the revisions for the manuscript were merged into main in #180

danbryce commented:
I rebased on main without that jetc-revert branch and pushed remove-jetc-revert.  I want to merge this new branch to main, and will delete the one attached to this PR.  Sound okay?

> On Mar 6, 2023, at 11:25 AM, bbartley ***@***.***> wrote:
> 
> 
> I don't think we need the jetc-revert branch. That branch should be pruned. I can't recall the problem I was trying to fix, but I think that I ended up resolving it without reverting.
> Looking back, the revisions for the manuscript were merged into main in #180
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.
>

bbartley commented:
Good to go!

danbryce commented:
Obviated by #187 (rebased on main)
